### ISSUE TITLE
Seems to ignore timezones

### ISSUE THREAD

**Issue by anentropic** [simon-weber/python-libfaketime#1]
I am passing in strings which `dateutil` docs say should result in a timezone aware datetime being returned, eg `@fake_time('7 Jul 2015 00:00:00 -0700')`

in my test I get:

```
ipdb> datetime.utcnow()
datetime.datetime(2015, 7, 6, 23, 0)
```

i.e. it has parsed the date string, discarded the tzinfo which dateutil should have parsed, then returned '7 Jul 2015 00:00:00' in my _local_ timezone as UTC

It appears the tzinfo is being discarded by code in `fake_time` decorator

**Comment by simon-weber** [comment]
Whoops; I don't have any datetimes that are timezone-aware in my tests.

Will everything just work if I parse into a timezone-aware datetime? I'm not sure offhand how that will interact with passing a string to libfaketime.

**Comment by anentropic** [comment]
I started trying to fix it:
https://github.com/anentropic/python-libfaketime/tree/preserve-timezone

I'm not having much luck, not sure if libfaketime itself really respects timezones

**Comment by simon-weber** [comment]
I ship with a fork of libfaketime, in case there's something we could do on that end to make things easier. Here's where the time gets parsed in: https://github.com/simon-weber/libfaketime/blob/python-libfaketime/src/libfaketime.c#L1477.

**Comment by anentropic** [comment]
C is not my thing but looking at the libfaketime code it seems like they just use a bare `strptime` call

googling for example code of C `strptime` and timezones it looks like additional steps may be necessary to work with `strptime` result in a timezone-aware fashion

eg
http://stackoverflow.com/a/1620451/202168

**Comment by anentropic** [comment]
bit out of my depth with the C stuff, if you could fix that it would be awesome

**Comment by simon-weber** [comment]
Haha, I'm no expert either, but I'll take a look when I've got a chance.

**Comment by anentropic** [comment]
currently using a workaround:

```python
import pytz
from django.conf import settings
from libfaketime import fake_time


def prepare_fake_time(target_datetime):
    """
    takes a timezone-aware datetime
    returns a naive datetime implicitly converted to local timezone
    """
    naive = target_datetime.replace(tzinfo=None)
    naive_utc = naive - target_datetime.utcoffset()
    # obviously this is django-specific way to get current timezone: 
    djtz = pytz.timezone(settings.TIME_ZONE)
    return djtz.fromutc(naive_utc).replace(tzinfo=None)


@fake_time(prepare_fake_time(tz_aware_datetime))
def test_something(self):
    pass
```

**Comment by adamchainz** [comment]
+1 In converting a large django test suite from `freeze_time` to `python-libfaketime` for speed purposes, we found this, and have resorted to a wrapper.

**Comment by gertjanol** [comment]
I'm using this workaround now, based on @anentropic's idea, for timezone aware patching, without the dependency on Django. It also makes sure it is a drop-in replacement for freezegun's `freeze_time`, so I only have to change my imports to make all existing tests pass.

```python
import datetime
import pytz
from dateutil import parser, tz
from libfaketime import fake_time

class freeze_time(fake_time):

    def __init__(self, spec, tz_offset=None, **kwargs):
        dt = self._prepare(spec, tz_offset=tz_offset)
        super(freeze_time, self).__init__(dt, **kwargs)

    def _prepare(self, spec, tz_offset=None):
        dt = spec if isinstance(spec, datetime.datetime) else parser.parse(spec)

        # If datetime currently has tzinfo, represent it as a datetime in UTC
        utc = self._convert_to_utc(dt)

        # If a tz_offset was given, subtract that also from the datetime
        if tz_offset is not None:
            utc -= datetime.timedelta(hours=tz_offset)

        # Convert the UTC datetime to the local timezone, and remove tzinfo,
        local = utc.astimezone(tz.tzlocal())
        return local.replace(tzinfo=None)

    @staticmethod
    def _convert_to_utc(dt):
        if dt.tzinfo:
            return dt.astimezone(pytz.utc)
        else:
            return dt.replace(tzinfo=pytz.utc)
```

**Comment by simon-weber** [comment]
@gertjanol nice! Want to open a PR for that? I'd love to add timezone support.

**Comment by gertjanol** [comment]
Sure, I'll see if I can whip something up. Do you care about the `freeze_time` interface that is compatible with freezegun, or do you prefer to keep `fake_time` as it is? Or both?

**Comment by simon-weber** [comment]
Both is probably ideal. Mainly I'd like to keep fake_time for backwards compatibility, but I'm fine with adding freeze_time, too.

**Comment by simon-weber** [comment]
I think this was fixed by #36.
